crypto.go

```diff
@@ -22,6 +22,52 @@ const (
 // the output of a SHA256 hash.
 type Hash256 [sha256.Size]byte
 
+// SingleKeyECDH is an abstraction interface that hides the implementation of an
+// ECDH operation against a specific private key. We use this abstraction for
+// the long term keys which we eventually want to be able to keep in a hardware
+// wallet or HSM.
+type SingleKeyECDH interface {
+	// PubKey returns the public key of the wrapped, "local" private key.
+	PubKey() *btcec.PublicKey
+
+	// ECDH performs a scalar multiplication (ECDH-like operation) between
+	// the wrapped private key and remote public key. The output returned
+	// will be the sha256 of the resulting shared point serialized in
+	// compressed format.
+	ECDH(pubKey *btcec.PublicKey) ([32]byte, error)
+}
+
+// PrivKeyECDH is an implementation of the SingleKeyECDH in which we do have the
+// full private key. This can be used to wrap a temporary key to conform to the
+// SingleKeyECDH interface.
+type PrivKeyECDH struct {
+	// PrivKey is the private key that is used for the ECDH operation.
+	PrivKey *btcec.PrivateKey
+}
+
+// PubKey returns the public key of the wrapped, "local" private key.
+//
+// NOTE: This is part of the SingleKeyECDH interface.
+func (p *PrivKeyECDH) PubKey() *btcec.PublicKey {
+	return p.PrivKey.PubKey()
+}
+
+// ECDH performs a scalar multiplication (ECDH-like operation) between the
+// wrapped private key and remote public key. The output returned will be the
+// sha256 of the resulting shared point serialized in compressed format. If k is
+// our private key, and P is the public key, we perform the following operation:
+//
+//  sx := k*P
+//  s := sha256(sx.SerializeCompressed())
+//
+// NOTE: This is part of the SingleKeyECDH interface.
+func (p *PrivKeyECDH) ECDH(pub *btcec.PublicKey) ([32]byte, error) {
+	s := &btcec.PublicKey{}
```

Comment:
Use the `generateSharedSecret` method here just so we have all the ECDH code in one place?

Reply:
A later commit will change the code of `generateSharedSecret` to just call this method instead. So the code will be _moved_ here in effect. I added a commit message to explain this.